posts/ast_primer.md

```diff
@@ -0,0 +1,405 @@
+]
+
+for table in tables:
+    print(table)
+
+# y  -- Success!
+```
+
+`build_scope` returns an instance of the `Scope` class. `Scope` has numerous methods for inspecting a query. The best way to browse these methods is directly in the code at [scope.py](../sqlglot/optimizer/scope.py). You can also look for examples of how Scope is used throughout SQLGlot's [optimizer](../sqlglot/optimizer) module. 
+
+Many methods of Scope depend on a fully qualified SQL expression. For example, let's say we want to trace the lineage of columns in this query:
+```python
+ast = parse_one("""
+SELECT
+  a,
+  c
+FROM (
+  SELECT 
+    a, 
+    b 
+  FROM x
+) AS x
+JOIN (
+  SELECT 
+    b, 
+    c 
+  FROM y
+) AS y
+  ON x.b = y.b
+""")
+```
+
+Just looking at the outer query, it's not obvious that column `a` comes from table `x` without looking at the columns of the subqueries.
+
+We can use the [qualify](../sqlglot/optimizer/qualify.py) function to prefix all columns in an AST with their table name like so:
+```python
+from sqlglot.optimizer.qualify import qualify
+
+qualify(ast)
+# SELECT
+#   x.a AS a,
+#   y.c AS c
+# FROM (
+#   SELECT
+#     x.a AS a,
+#     x.b AS b
+#   FROM x AS x
+# ) AS x
+# JOIN (
+#   SELECT
+#     y.b AS b,
+#     y.c AS c
+#   FROM y AS y
+# ) AS y
+#   ON x.b = y.b
+```
+
+Now we can trace a column to its source. Here's how we might find the table or subquery for all columns in a qualified AST:
+```python
+from sqlglot.optimizer.scope import find_all_in_scope
+
+root = build_scope(ast)
+
+# `find_all_in_scope` is similar to `Expression.find_all`, except it doesn't traverse into subqueries
+for column in find_all_in_scope(root.expression, exp.Column):
+    print(f"{column} => {root.sources[column.table]}")
+
+# x.a => Scope<SELECT x.a AS a, x.b AS b FROM x AS x>
+# y.c => Scope<SELECT y.b AS b, y.c AS c FROM y AS y>
+# x.b => Scope<SELECT x.a AS a, x.b AS b FROM x AS x>
+# y.b => Scope<SELECT y.b AS b, y.c AS c FROM y AS y>
+```
+
+For a complete example of tracing column lineage, check out the [lineage](../sqlglot/lineage.py) module.
+
+> [!NOTE]
+> Some queries require the database schema for disambiguation. For example:
+> 
+> ```sql
+> SELECT a FROM x CROSS JOIN y
+> ```
+> 
+> Column `a` might come from table `x` or `y`. In these cases, you must pass the `schema` into `qualify`.
+
+## Mutating the tree
+
+You can also modify an AST or build one from scratch. There are a few ways to do this.
+
+### High-level builder methods
+
+SQLGlot has methods for programmatically building up expressions similar to how you might in an ORM:
+```python
+ast = (
+    exp
+    .select("a", "b")
+    .from_("x")
+    .where("b < 4")
+    .limit(10)
+)
+```
+
+> [!WARNING]
+> High-level builder methods will attempt to parse string arguments into Expressions. This can be very convenient, but make sure to keep in mind dialect of the string. If its written in a specific dialect, you need to set the `dialect` argument.
```

Comment:
```suggestion
> High-level builder methods will attempt to parse string arguments into Expressions. This can be very convenient, but make sure to keep in mind the dialect of the string. If its written in a specific dialect, you need to set the `dialect` argument.
```